libPhoneNumber/NBPhoneNumberOfflineGeocoder.m

```diff
@@ -0,0 +1,173 @@
+  NSArray *filesInResourceBundle =
+      [[NSFileManager defaultManager] contentsOfDirectoryAtURL:bundleURL
+                                    includingPropertiesForKeys:NULL
+                                                       options:0
+                                                         error:NULL];
+  NSMutableArray *supportedLanguages = [[NSMutableArray alloc] init];
+  [filesInResourceBundle enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
+    NSString *filename = (NSString *)obj;
+    if (filename != NULL) {
+      [supportedLanguages addObject:[[filename lastPathComponent] stringByDeletingPathExtension]];
+    }
+  }];
+  NSString *language;
+  // need to get from ios device settings
+  if ([[NSLocale preferredLanguages] count] > 0) {
+    // set language to preferred
+    language = [NSLocale preferredLanguages][0];
+    // if the preferred language isn't supported by metadata, set by default to english
+    if (![supportedLanguages containsObject:language]) {
+      language = @"en";
+      NSLog(@"Set default language to ENGLISH");
+    }
+  } else {
+    language = @"en";
+  }
+
+  self->geocoderHelper = [[NBGeocoderMetadataHelper alloc] initWithCountryCode:@1
+                                                                  withLanguage:language];
+  return self;
+}
+
+- (NSString *)countryNameForNumber:(NBPhoneNumber *)number
+                  withLanguageCode:(NSString *)languageCode {
+  if (number == NULL) return @"";
+  NSArray *regionCodes = [self->phoneUtil getRegionCodesForCountryCode:number.countryCode];
+  if ([regionCodes count] == 1) {
+    return [self regionDisplayName:[regionCodes objectAtIndex:0] withLanguageCode:languageCode];
+  } else {
+    NSString *regionWhereNumberIsValid = @"ZZ";
+    for (NSString *regionCode in regionCodes) {
+      if ([self->phoneUtil isValidNumberForRegion:number regionCode:regionCode]) {
+        if (![regionWhereNumberIsValid isEqualToString:@"ZZ"]) {
+          NSLog(@"multiple valid regions found, so returning none");
+          return @"";
+        }
+        NSLog(@"Setting regionWhereNumberIsValid to %@", regionCode);
+        regionWhereNumberIsValid = regionCode;
+      }
+    }
+
+    return [self regionDisplayName:regionWhereNumberIsValid withLanguageCode:languageCode];
+  }
+}
+
+- (NSString *)regionDisplayName:(NSString *)regionCode withLanguageCode:(NSString *)languageCode {
+  return (regionCode == NULL || [regionCode isEqualToString:@"ZZ"] ||
+          [regionCode isEqual:NB_REGION_CODE_FOR_NON_GEO_ENTITY])
+             ? @""
+             : [[[NSLocale alloc] initWithLocaleIdentifier:languageCode]
+                   displayNameForKey:NSLocaleCountryCode
+                               value:regionCode];
+}
+
+- (NSString *)descriptionForValidNumber:(NBPhoneNumber *)number
+                       withLanguageCode:(NSString *)languageCode {
+  [self->geocoderHelper setLanguage:languageCode];
```

Comment:
Instead of manipulating geocoderHelper, create a new instance of the helper, put it in the cache (if not exists), and if the helper for that particular language already exists, then just grab the instance out of the cache.

Reply:
Completed!